File: leetcode/374_guess_number_higher_lower.py
# We are playing the Guess Game. The game is as follows:
#
# I pick a number from 1 to n. You have to guess which number I picked.
#
# Every time you guess wrong, I will tell you whether the number I picked is higher or lower than your guess.
#
# You call a pre-defined API int guess(int num), which returns 3 possible results:
#
# -1: The number I picked is lower than your guess (i.e. pick < num).
# 1: The number I picked is higher than your guess (i.e. pick > num).
# 0: The number I picked is equal to your guess (i.e. pick == num).
# Return the number that I picked.
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        low = 1
        high = n

        # implement binary search
        while (low <= high):
            mid = math.floor(low + (high - low) / 2)
            res = guess(mid)

            if (res == 0):
                return mid
            # number is lower than the guess therefore move our high to the guess(mid) - 1
            elif (res < 0):
                high = mid - 1
            # number is higher than the guess therefore move our low to the guess(mid) + 1
            else:
                low = mid + 1

        return -1
# Time: O(log N)
# Space: O(1)

File: leetcode/374_guess_number_higher_lower.js
// We are playing the Guess Game. The game is as follows:
//
// I pick a number from 1 to n. You have to guess which number I picked.
//
// Every time you guess wrong, I will tell you whether the number I picked is higher or lower than your guess.
//
// You call a pre-defined API int guess(int num), which returns three possible results:
//
// -1: Your guess is higher than the number I picked (i.e. num > pick).
// 1: Your guess is lower than the number I picked (i.e. num < pick).
// 0: your guess is equal to the number I picked (i.e. num == pick).
// Return the number that I picked.

/**
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * var guess = function(num) {}
 */

/**
 * @param {number} n
 * @return {number}
 */
var guessNumber = function (n) {
  let low = 1;
  let high = n;

  while (low <= high) {
    let mid = Math.floor(low + (high - low) / 2);
    let res = guess(mid);

    if (res == 0) {
      return mid;
      // guess is lower than the target
      // then our number should be between the mid and high
    } else if (res == 1) {
      low = mid + 1;
      // guess is higher than the target
      // then our number should be between the low and mid
    } else {
      high = mid - 1;
    }
  }

  return 0;
};
